Reflection variable resolver

kpavlov opened the pull request:
## Issue
Cleaning up to prepare for implementing async AI Services

## Change
- This PR refactors and improves how method argument values are extracted and mapped for prompt templates in AI service methods. It moves variable resolution logic out of the main service implementation and into a dedicated utility class, which enhances code clarity, modularity, and reusability. **The new utility also safely handles null argument arrays, preventing possible NPE errors.** Minor code formatting and organization tweaks were also included for better readability and consistency. No feature or behavioral changes were introduced.
- Enabled OpenAI tests to conditionally run based on the presence of the `OPENAI_API_KEY` environment variable. This ensures tests are only executed when the necessary API key is available, improving reliability in controlled environments.
- Add unit tests for ReflectionVariableResolver

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

kpavlov committed:
Refactor variable resolution into ReflectionVariableResolver

Extracted variable resolution logic from DefaultAiServices to a new utility class, ReflectionVariableResolver. This improves code modularity, reusability, and readability by centralizing the reflection-based variable handling logic.

kpavlov committed:
`Handle null arguments in ReflectionVariableResolver`

Ensure `findTemplateVariables` gracefully handles null `args` by returning an empty map. Adjust iteration logic to prevent potential issues when `args` is null.

kpavlov committed:
Reorganized static imports

Reorganized static imports and reformatted code to enhance clarity and maintainability. Adjusted method formatting and indentation for better consistency and compliance with coding standards. No functional changes introduced.

kpavlov committed:
Add environment variable checks to integration tests

Enabled tests to conditionally run based on the presence of the `OPENAI_API_KEY` environment variable. This ensures tests are only executed when the necessary API key is available, improving reliability in controlled environments. Additionally, minor formatting adjustments were applied for improved code readability.

kpavlov committed:
Handle null argument in ReflectionVariableResolver

Previously, passing a null value to the asString method caused unexpected behavior. This update adds explicit handling for null values, returning "null" as a string.

kpavlov committed:
Add unit tests for ReflectionVariableResolver

Introduce comprehensive unit tests to validate the behavior of `ReflectionVariableResolver`. These tests cover various scenarios, including handling annotations, single and array parameters, null arguments, and template variable mapping.

mariofusco commented on langchain4j/src/main/java/dev/langchain4j/service/ReflectionVariableResolver.java:
```diff
@@ -0,0 +1,106 @@
+ * This class is intended for internal use only and is designed to extract and map
+ * parameter values to template variables in methods defined within AI services.
+ */
+public class ReflectionVariableResolver {
```
Could this be package private?

kpavlov replied:
I love weaker access modifiers, but want to make it open for reuse by extensions, e.g. in [my AsyncAiServices ](https://github.com/kpavlov/langchain4j-kotlin/pull/128). It's shaping up

dliubarskyi replied:
Then let's annotate it `@Internal` and rename into something like `InternalReflectionVariableResolver`

mariofusco reviewed (dismissed):
Nice refactor, thank you. Let's just try to keep all classes and methods with the most restricting visibility possible. In general we need to reduce the surface of all our public APIs.

kpavlov commented:
> Nice refactor, thank you. Let's just try to keep all classes and methods with the most restricting visibility possible. In general we need to reduce the surface of all our public APIs.

+1, but I want to use it Kotlin AIServices, so, I think it should be public.

kpavlov committed:
Merge branch 'main' into reflection-resolver

dliubarskyi committed:
Merge branch 'main' into reflection-resolver

dliubarskyi reviewed (commented):
@kpavlov thank you, nice refactoring

kpavlov committed:
Merge branch 'main' into reflection-resolver

kpavlov committed:
Rename ReflectionVariableResolver to InternalReflectionVariableResolver

Updated the class name and references to better indicate internal usage. Added @Internal annotation.

dliubarskyi reviewed (approved):
@kpavlov thank you!
